### PR TITLE
Fix Retrofit template with Jackson for Kotlin

### LINKED ISSUE #8147: [BUG] [Kotlin] jvm-retrofit2 with Jackson issues
#### Bug Report Checklist

- [ ] Have you provided a full/minimal spec to reproduce the issue?
- [ ] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

When generating kotlin retrofit client with Jackson, the ApiClient doesn't compile and other small issues.

Here's what I have:
```kotlin
class ApiClient(
    private var baseUrl: String = defaultBasePath,
    private val okHttpClientBuilder: OkHttpClient.Builder? = null,
    private val serializerBuilder: Builder = Serializer.Builder, // Builder is unresolved, serializerBuilder isn't used in the class
    private val okHttpClient: OkHttpClient? = null
) { 
...
    // in my spec I have 3 requirements: basic, jwt, and oauth2. So, one of these should be fulfilled to access the api
    fun setCredentials(username: String, password: String): ApiClient {
        apiAuthorizations.values.runOnFirst<Interceptor, HttpBasicAuth> {
            setCredentials(username, password)
        }
        apiAuthorizations.values.runOnFirst<Interceptor, OAuth> {
            tokenRequestBuilder.setUsername(username).setPassword(password)
        }
        return this
    }

    // this method should not be here, it's already handled above in the setCredentials
    // this method should only be present when the is oauth but there is no basic auth
    fun setCredentials(username: String, password: String): ApiClient { 
        apiAuthorizations.values.runOnFirst<Interceptor, OAuth> {
            tokenRequestBuilder.setUsername(username).setPassword(password)
        }
        return this
    }
...
}
```
Also, okhttp version collision, logging-interceptor wants 4.4.0, retrofit wants 3.14.9 (https://github.com/square/retrofit/issues/3384). So, given I fix the interceptor version to 3.14.9 as well, I still have to fix some issues. 
`ApiKeyAuth.kt` doesn't compile:
```kotlin
...
if ("query" == location) {
    var newQuery = request.url.toUri().query // request.url does not exist, request.url() returns okhttp3.HttpUrl
    val paramValue = "$paramName=$apiKey"
    if (newQuery == null) {
        newQuery = paramValue
    } else {
        newQuery += "&$paramValue"
    }

    val newUri: URI
    try {
        val oldUri = request.url.toUri()
        newUri = URI(
            oldUri.scheme, oldUri.authority,
            oldUri.path, newQuery, oldUri.fragment
        )
    } catch (e: URISyntaxException) {
        throw IOException(e)
    }

    request = request.newBuilder().url(newUri.toURL()).build()
} else ...
...
```
Here's how it should be (I think):
```kotlin
if ("query" == location) {
    val newUrl = request.url().newBuilder().addQueryParameter(paramName, apiKey).build()
    request = request.newBuilder().url(newUrl).build()
} else ...
```
Also, `OAuth.kt` doesn't compile because it uses `request.url` and `request.body` instead of `request.url()` and `request.body()`.
Also `OAuthOkHttpClient.kt` doesn't compile because of `request.body` and because it uses a non-existent extension `toMediaTypeOrNull()` which I think can be replaced by static call `MediaType.parse(...)`.
Also, `ApiClient#retrofitBuilder` should really be public to allow customizations. 
Also, maybe `ApiClient` should be `open` for extension? I can extend it if generated by Java.
Also, there are empty files which only contain a package: `ByteArrayAdapter.kt`, `ResponseExt.kt`.
Also, there is a `Serializer.kt`which has a method but it's not used either:
```kotlin
object Serializer {
    @JvmStatic
    val jacksonObjectMapper: ObjectMapper = jacksonObjectMapper()
        .registerModule(Jdk8Module())
        .registerModule(JavaTimeModule())
        .setSerializationInclusion(JsonInclude.Include.NON_ABSENT)
        .configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false)
}
```
Also, the resulting code cannot serialize JSON, so I had to add the dependency`compile "com.squareup.retrofit2:converter-jackson:$retrofitVersion"` and this line to the `ApiClient`: 
```kotlin
private val retrofitBuilder: Retrofit.Builder by lazy {
    Retrofit.Builder()
        .baseUrl(baseUrl)
        .addConverterFactory(ScalarsConverterFactory.create())
        .addConverterFactory(JacksonConverterFactory.create(Serializer.jacksonObjectMapper)) // I added this line
}
```
Also, the markdown docs for each API class seem to document all authentication methods at the same time (but oauth is missing):
```kotlin
val apiClient = ApiClient()
apiClient.setCredentials("USERNAME", "PASSWORD")
apiClient.setBearerToken("TOKEN")
val webService = apiClient.createWebservice(...
```
Also, logging interceptor doesn't do anything by default. Could be refactoring like this (default constructor take care of it):
```kotlin
.addInterceptor(
    (if (logger != null) HttpLoggingInterceptor(logger) else HttpLoggingInterceptor()).apply {
        level = HttpLoggingInterceptor.Level.BODY
    }
)
```
Also, `ApiClient` accepts `authName` in specialized constructor even though it is redundant. For example, it asks for username and password, so it already knows that the authentication method is basic and there is no need to explicitly ask for its name.
Also, method `ApiClient#createService` accepts java class but it should accept kotlin class for convenience. 
Also, this line contains multiple statements a looks weird, should be reformatted:
https://github.com/OpenAPITools/openapi-generator/blob/9e5610488f8b649856a72cd25695fc071b0ca1c4/modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-retrofit2/infrastructure/ApiClient.kt.mustache#L106

Links to mustache: [ApiClient](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-retrofit2/infrastructure/ApiClient.kt.mustache), [ResponseExt.kt](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-retrofit2/infrastructure/ResponseExt.kt.mustache), [Serializer.kt](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/kotlin-client/jvm-common/infrastructure/Serializer.kt.mustache), [ByteArrayAdapter.kt](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/kotlin-client/jvm-common/infrastructure/ByteArrayAdapter.kt.mustache), [ApiKeyAuth.kt](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-retrofit2/auth/ApiKeyAuth.kt.mustache)

And I am not sure if this is relevant but on the Retrofit's homepage they say there is a Jackson converter at `com.squareup.retrofit2:converter-jackson` that we could use.

##### openapi-generator version

latest build from master

##### Steps to reproduce

```bash
openapi-generator generate \
  --generator-name kotlin \
  ... 
  --library jvm-retrofit2 \
  --additional-properties serializationLibrary=jackson
```
the `...` is ambiguous but I think the only thing that matters here is that it's a combination of retrofit and Jackson but let me know.

### PR DESCRIPTION
**Pull request by yonatankarp**
To fix #8147

This commit amends the `ApiClient` template for Retrofit. It fixes a bug where the Retrofit client uses Jackson as the serialization library.

Until now, the `ApiClient` could not be compiled when using the Jackson library (default in SpringBoot applications), as the constructor of the class had the following parameter:

```kotlin
private val serializerBuilder: Builder = Serializer.Builder,
```

You can see it more clearly here:

image

After the fix the outcome of this line would be:

```kotlin
private val serializerBuilder: ObjectMapper = Serializer.jacksonObjectMapper,
```

The configuration that was used for the test is as follows:

```kotlin
config = mapOf(
    "dateLibrary" to "java8",
    "interfaceOnly" to "true",
    "implicitHeaders" to "true",
    "hideGenerationTimestamp" to "true",
    "useTags" to "true",
    "documentationProvider" to "none",
    "serializationLibrary" to "jackson",
    "useCoroutines" to "true",
    "library" to "jvm-retrofit2"
)
```

**Commits:**

1. The bug fix itself
2. Adding code sample for CI comparison
3. Update existing samples with the new change.

All commits would be squashed into a single commit when this PR is approved, and are separated only for easier review process.

---

The change done in the commit is to ensure that in the case of Jackson, the right property of the `Serializer` class is assigned

validation of this change was by running the generator and ensuring that code the generated code passed compilation.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.3.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Review by yonatankarp — COMMENTED** [review]
@jimschubert could you please have a look? Noticed this bug at work yesterday. We've fixed it locally by altering the templates, but I think that it's better to fix it on the generator level.

**Review by yonatankarp — COMMENTED** [review]
@jimschubert after rebasing again on `master` all broken CI's have passed. can you please review this small fix? 🙏

**Comment by wing328** [comment]
@yonatankarp thanks for the PR. Can you please PM me via Slack to discuss this further?

https://join.slack.com/t/openapi-generator/shared_invite/zt-12jxxd7p2-XUeQM~4pzsU9x~eGLQqX2g

File: bin/configs/kotlin-jvm-jackson.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+generatorName: kotlin
+outputDir: samples/client/petstore/kotlin-jvm-jackson
+inputSpec: modules/openapi-generator/src/test/resources/2_0/petstore.yaml
```
Comment (wing328) [inline bin/configs/kotlin-jvm-jackson.yaml:3]: I did a test locally but got the following errors:
```
➜  kotlin-jvm-jackson git:(fix_retrofit_generator_with_jackson) ✗ gradle build
Starting a Gradle Daemon (subsequent builds will be faster)

> Task :compileKotlin
e: /Users/williamcheng/gitclone/yonatankarp_fix_retrofit_generator_with_jackson/samples/client/petstore/kotlin-jvm-jackson/src/main/kotlin/org/openapitools/client/infrastructure/ApiClient.kt: (60, 28): Unresolved reference: Builder
e: /Users/williamcheng/gitclone/yonatankarp_fix_retrofit_generator_with_jackson/samples/client/petstore/kotlin-jvm-jackson/src/main/kotlin/org/openapitools/client/infrastructure/ApiClient.kt: (60, 49): Unresolved reference: Builder
e: /Users/williamcheng/gitclone/yonatankarp_fix_retrofit_generator_with_jackson/samples/client/petstore/kotlin-jvm-jackson/src/main/kotlin/org/openapitools/client/infrastructure/ApiClient.kt: (75, 28): Unresolved reference: Builder
e: /Users/williamcheng/gitclone/yonatankarp_fix_retrofit_generator_with_jackson/samples/client/petstore/kotlin-jvm-jackson/src/main/kotlin/org/openapitools/client/infrastructure/ApiClient.kt: (75, 49): Unresolved reference: Builder

> Task :compileKotlin FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileKotlin'.
> A failure occurred while executing org.jetbrains.kotlin.compilerRunner.GradleCompilerRunnerWithWorkers$GradleKotlinCompilerWorkAction
   > Compilation error. See log for more details

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

```

Can you please take a look when you've time?

Also update use 3.0 spec instead: `modules/openapi-generator/src/test/resources/3_0/petstore.yaml`

Comment (wing328) [reply bin/configs/kotlin-jvm-jackson.yaml:3]: Please commit the new sample files as well so that github workflow will trigger the build.

Comment (yonatankarp) [reply bin/configs/kotlin-jvm-jackson.yaml]: There was indeed a but with my fix - my bad! 🤦‍♂️ 
I have fixed it and updated the sample app and the spec version however seems that `GitHub Actions` isn't picking the changes yet, and `CircleCI` still failing after I have rebased on `master`

**Comment by wing328** [comment]
please ignore the circleci failure for the time being.

can you please update the samples? 

ref: https://github.com/OpenAPITools/openapi-generator/actions/runs/5840007126/job/15839427531?pr=14239

**Review by yonatankarp — COMMENTED** [review]
@wing328 I have rebased the branch on `master` again (seems that previously it was wrongly done as the generator version was `6.3.0-SNAPSHOT` on my branch) and also executed the command:

```bash
$ bin/generate-samples.sh bin/configs/kotlin-*
```

Everything should (hopefully) be up-to-date now, and waiting for an approval for GitHub to run the pipeline

**Comment by wing328** [comment]
Thanks for the PR but your commit (as shown in the Commits tab) is not linked to your Github account, which means this PR won't count as your contribution in https://github.com/OpenAPITools/openapi-generator/graphs/contributors.

Let me know if you need help fixing it.

Ref: https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-can-i-update-commits-that-are-not-linked-to-my-github-account

**Comment by wing328** [comment]
https://github.com/OpenAPITools/openapi-generator/actions/runs/5847743068/job/15854622468?pr=14239 failure can be ignored for the time being.

**Review by yonatankarp — COMMENTED** [review]
Thanks a lot for noticing that!

I believe that now the commits are linked to your GitHub account.

**Comment by wing328** [comment]
tests passed: https://github.com/OpenAPITools/openapi-generator/actions/runs/5848309681/job/15860965534?pr=14239

**Pull request merged**